### PR TITLE
Get free Variables

### PR DESCRIPTION
**Pull request by marcogario**
Refactoring of FNode.get_dependencies() to use iterative walkers.

- Introduces the FreeVarsOracle, and the get_free_variables function/shortcut.
- Added Deprecation warning to get_dependencies
- Updated code-base to use get_free_variables

### CODE REVIEW

File: pysmt/oracles.py [hunk 1]
```diff
@@ -168,10 +169,73 @@ def get_theory(self, formula):
...
+
+    def walk_symbol(self, formula, args):
+        return frozenset([formula])
+
+    def walk_constant(self, formula, args):
+        return frozenset()
+
+    def walk_function(self, formula, args):
+        # MG: Do we need to create a list?
+        res = set([formula._content.payload])
```
Comment (mikand) [inline pysmt/oracles.py]: I would avoid using _content.
This is equivalent to ```formula.function_name()```

File: pysmt/oracles.py [hunk 2]
```diff
@@ -168,10 +169,73 @@ def get_theory(self, formula):
...
+        return self.walk(formula)
+
+    def walk_simple_args(self, formula, args):
+        res = set()
+        for arg in args:
+            res.update(arg)
+        return frozenset(res)
+
+    def walk_quantifier(self, formula, args):
+        return args[0].difference(formula._content.payload)
```
Comment (mikand) [inline pysmt/oracles.py]: I would avoid using _content.
This is equivalent to formula.quantifier_vars()

File: pysmt/oracles.py [hunk 3]
```diff
@@ -168,10 +169,73 @@ def get_theory(self, formula):
...
 
-
 # EOC TheoryOracle
 
 
+# Operators for which Args is an FNode
+DEPENDENCIES_SIMPLE_ARGS = (set(op.ALL_TYPES) - \
+                           (set([op.SYMBOL, op.FUNCTION]) | op.QUANTIFIERS | op.CONSTANTS))
+
+class FreeVarsOracle(pysmt.walkers.DagWalker):
```
Comment (mikand) [inline pysmt/oracles.py]: A possible improvement (useful for function models) could be to have a parameter to control if/how function sysmbols are added in the resulting cone.

I see three modes:
* None: meaning that only variables are considered in the cone
* Function Symbols [default and current behavior]: function names are symbols appearing in the cpone, but function applications are discarded
* Function applications: function applications are considered atomic and are retured together with the cones of theit parameters.

Comment (marcogario) [inline pysmt/oracles.py]: Lets keep this for a future. I think this goes in the direction of having an implementation in which we do Free Variables vs All Variables. Lets open a ticket with all "versions" that we want to have.

**Pull request merged**